### PR TITLE
Add stable diffusion unit test

### PR DESCRIPTION
**Pull request by mrwyattii**
Add a unit test for Stable Diffusion without using stable-diffusion model that needs HF token. 

Midjourney model does not need the HF token and has the same structure.

We verified this by printing both models. The following is the structure:

``` StableDiffusionPipeline {
  "_class_name": "StableDiffusionPipeline",
  "_diffusers_version": "0.7.2",
  "feature_extractor": [
    "transformers",
    "CLIPFeatureExtractor"
  ],
  "safety_checker": [
    "stable_diffusion",
    "StableDiffusionSafetyChecker"
  ],
  "scheduler": [
    "diffusers",
    "PNDMScheduler"
  ],
  "text_encoder": [
    "transformers",
    "CLIPTextModel"
  ],
  "tokenizer": [
    "transformers",
    "CLIPTokenizer"
  ],
  "unet": [
    "diffusers",
    "UNet2DConditionModel"
  ],
  "vae": [
    "diffusers",
    "AutoencoderKL"
  ]
}
```
@cmikeh2

### CODE REVIEW

**Review by awan-10 — APPROVED** [review]

**Comment by Thomas-MMJ** [comment]
You can set a manual seed, and the images should be the same,

```
generator = torch.Generator("cuda").manual_seed(1024)
image = pipe(prompt, guidance_scale=7.5, generator=generator).images[0]
```

https://huggingface.co/blog/stable_diffusion

File: tests/unit/inference/test_inference.py [hunk 1]
```diff
@@ -341,6 +341,46 @@ def test(
         assert assert_fn(bs_output, ds_output)
 
 
+# Setup for these models is different from other pipelines, so we add a separate test
+@pytest.mark.inference
```
Comment (mrwyattii) [inline tests/unit/inference/test_inference.py:345]: We don't want this running on `inference` workflows, let's make a new `pytest.mark.stable_diffusion`.

Can we also move the test to a new file `test_stable_diffusion.py`?

File: tests/unit/inference/test_inference.py [hunk 2]
```diff
@@ -341,6 +341,46 @@ def test(
...
+class TestStableDiffusion(DistributedTest):
+    world_size = 1
+
+    def test(self):
+        from diffusers import DiffusionPipeline
+        from image_similarity_measures.evaluate import evaluation
```
Comment (mrwyattii) [inline tests/unit/inference/test_inference.py:351]: Make sure to add this to the requirements somewhere

File: .github/workflows/nv-sd.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+name: nv-a6000
```
Comment (loadams) [inline .github/workflows/nv-sd.yml:1]: You'll want to update the name here

File: .github/workflows/nv-sd.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+permissions:
+  contents: read
+  issues: write
+
+jobs:
+  unit-tests:
+    runs-on: [self-hosted, nvidia, a6000]
```
Comment (loadams) [inline .github/workflows/nv-sd.yml:18]: Not sure what hardware we want this running on, but specify here?

File: .github/workflows/nv-sd.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    container:
+      image: nvcr.io/nvidia/pytorch:23.03-py3
```
Comment (loadams) [inline .github/workflows/nv-sd.yml:20]: And depending on the hardware, this container section may need to go?

File: .github/workflows/nv-sd.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+name: nv-a6000
+
+on:
+  schedule:
+    - cron: "0 0 * * 0"
+  workflow_dispatch:
```
Comment (mrwyattii) [inline .github/workflows/nv-sd.yml:6]:
```suggestion
  schedule:
    - cron: "0 0 * * 0"
  workflow_dispatch:
  pull_request:
    paths:
      - "deepspeed/ops/transformer/inference/diffusers_**"
      - "tests/unit/inference/test_stable_difussion.py"
```

Something like this should make the test trigger anytime there are changes to Stable Diffusion code/tests (but there may be additional files to add)

This should make the test trigger

File: tests/unit/inference/test_inference.py (hunk 2 above)
Comment (loadams) [reply tests/unit/inference/test_inference.py:351]: Could add this to requirements-sd?

File: .github/workflows/nv-sd.yml (hunk 6 above)
Comment (lekurile) [reply .github/workflows/nv-sd.yml:6]: Updated, thanks.

File: .github/workflows/nv-sd.yml (hunk 3 above)
Comment (lekurile) [reply .github/workflows/nv-sd.yml]: Updated, thanks.

File: tests/unit/inference/test_inference.py (hunk 2 above)
Comment (lekurile) [reply tests/unit/inference/test_inference.py]: Added to requirements-sd, thanks.

File: tests/unit/inference/test_inference.py (hunk 1 above)
Comment (lekurile) [reply tests/unit/inference/test_inference.py]: Updated both items, thanks.

File: .github/workflows/nv-sd.yml (hunk 4 above)
Comment (lekurile) [reply .github/workflows/nv-sd.yml:22]: For stable diffusion we want to target A6000 due to Triton's architecture support.

**Review by lekurile — APPROVED** [review]

File: .github/workflows/nv-inference.yml [hunk 7]
```diff
@@ -39,6 +39,7 @@ jobs:
 
       - name: Install deepspeed
         run: |
+          pip install diffusers
```
[the remark this answers is not in the document]
Comment (lekurile) [reply .github/workflows/nv-inference.yml]: Updated.

File: .github/workflows/nv-sd.yml (hunk 5 above)
Comment (loadams) [reply .github/workflows/nv-sd.yml:24]: Nevermind, we want to run on a6000

File: .github/workflows/nv-sd.yml [hunk 8]
```diff
@@ -55,6 +55,7 @@ jobs:
       - name: Unit tests
         run: |
           unset TORCH_CUDA_ARCH_LIST # only jit compile for current arch
+          python download_sd_model.py
```
Comment (loadams) [inline .github/workflows/nv-sd.yml:58]: I'd probably move this to tests/unit/utils or test/unit/inference?

Comment (lekurile) [reply .github/workflows/nv-sd.yml]: Script was removed, marking as resolved.

**Review by jeffra — APPROVED** [review]

File: .github/workflows/nv-sd.yml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+permissions:
+  contents: read
+  issues: write
+
+jobs:
+  unit-tests:
```
Comment (jeffra) [inline .github/workflows/nv-sd.yml:23]: can we rename this to something like `sd-tests`? then we won't require it to pass for all PRs even if they don't match the trigger path

Comment (lekurile) [reply .github/workflows/nv-sd.yml]: Updated in latest commit, thanks for the suggestion.

**Pull request merged**